### PR TITLE
vm/Tags

### PR DESCRIPTION
**Pull request by milanaghera-tenable**
This PR contains following change for Tags API endpoint:

- change in folder structure as per v3 standard
- added search method and remove corresponding list method
- added marshmallow schema validation for validating the request body.
- test case will be found under the `tests/io/v3/vm/tags` modules

### CODE REVIEW

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+    Test case for tags details method
+    '''
+    value_id: str = '60bfce80-f695-4ed0-bd47-ec9e5b2946e5'
+    test_response: dict = {
+        'id': value_id,
+        'created_at': '2021-12-30T05:34:46.626Z',
+        'created_by': 'jyoti.patel@crestdatasys.com',
+        'updated_at': '2021-12-30T05:34:46.626Z',
+        'updated_by': 'jyoti.patel@crestdatasys.com',
+        'category_uuid': '4c5f6aaf-22fe-458d-a7a3-76c60e90147a',
```
Comment (harshshah-tenable) [inline tests/io/v3/vm/tags/test_tags_api.py:146]: category_id

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+                'ALL',
+                'CAN_EDIT',
+                'CAN_SET_PERMISSIONS',
+                'CAN_USE'
+            ],
+            'all_users_permissions': [],
+            'current_domain_permissions': [
+                {
+                    'id': 'e9f23194-adb7-4c02-8632-615c694c787e',
+                    'name': 'jyoti.patel@crestdatasys.com',
```
Comment (harshshah-tenable) [inline tests/io/v3/vm/tags/test_tags_api.py:169]: update here and all other instances with some test user `user1@gmail.com`

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+                    ],
+                    'name': 'user@company.com',
+                    'type': 'USER',
+                    'id': 'c2f2d080-ac2b-4278-914b-29f148682ee1'
+                }
+            ]
+        }
+    }
+    test_response: dict = {
+        'uuid': '8df056ea-a569-46a7-a443-fb58b34f20e6',
```
Comment (harshshah-tenable) [inline tests/io/v3/vm/tags/test_tags_api.py:362]: id

File: tenable/io/v3/vm/tags/schema.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    name = fields.Str(required=True)
+    description = fields.Str()
+
+    @validates_schema
+    def validate_name(self, data, **kwargs):
+        '''
+        Supproting method to validate name string length
+        '''
+        if len(data['name']) >= 127:
+            raise ValidationError('The name must not exceed 127 characters.')
```
Comment (sahils-tenable) [inline tenable/io/v3/vm/tags/schema.py]: can we use any marshmallow validator ex= v.length(max=127) => pls check the marshmallow document https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html

File: tenable/io/v3/vm/tags/schema.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    '''
+    Supporting schema for TagValueSchema class
+    '''
+    current_user_permissions = fields.List(
+        fields.Str(
+            validate=validate.OneOf(
+                ['ALL', 'CAN_USE', 'CAN_EDIT', 'CAN_SET_PERMISSIONS']
+            )
+        ),
+        default=['ALL', 'CAN_EDIT', 'CAN_SET_PERMISSIONS'],
```
Comment (sahils-tenable) [inline tenable/io/v3/vm/tags/schema.py]: see if we can use load_default or dump_default, and applicable for everywhere possible

File: tenable/io/v3/vm/tags/schema.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        default=[]
+    )
+    defined_domain_permissions = fields.List(
+        fields.Str(
+            validate=validate.OneOf(
+                ['ALL', 'CAN_USE', 'CAN_EDIT', 'CAN_SET_PERMISSIONS']
+            )
+        ),
+    )
+    version = fields.Int(data_key='version')
```
Comment (sahils-tenable) [inline tenable/io/v3/vm/tags/schema.py]: both data_key and actual variable name is same. is it intentional?

File: tenable/io/v3/vm/tags/schema.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    '''
+    _filters = None
+
+
+class TagValueSchema(Schema):
+    '''
+    Schema for tags values related methods
+    '''
+    value_id = fields.List(fields.UUID, data_key='values')
+    category_id = fields.UUID(data_key='category_uuid')
```
Comment (sahils-tenable) [inline tenable/io/v3/vm/tags/schema.py]: do we need to convert this category_uuid to category_id?

File: tenable/io/v3/vm/tags/schema.py (hunk 4 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/tags/schema.py]: Resolved

File: tenable/io/v3/vm/tags/schema.py (hunk 5 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/tags/schema.py]: Resolved.

File: tenable/io/v3/vm/tags/schema.py (hunk 6 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/tags/schema.py]: Removed.

File: tenable/io/v3/vm/tags/schema.py (hunk 7 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/tags/schema.py]: Resolved.

**Review by dhvanilshah-tenable — APPROVED** [review]

**Review by sahils-tenable — APPROVED** [review]

**Review by parvdesai-tenable — APPROVED** [review]

File: tests/io/v3/vm/tags/test_tags_api.py (hunk 1 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/tags/test_tags_api.py]: done.

File: tests/io/v3/vm/tags/test_tags_api.py (hunk 2 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/tags/test_tags_api.py]: done.

File: tenable/io/v3/vm/tags/api.py [hunk 8]
```diff
@@ -0,0 +1,746 @@
...
+            ...     'Madison'
+            ... )
+        '''
+
+        # First lets see if the category is a UUID or a general string.  If its
+        # a UUID, then we will pass the value of category into the
+        # category_uuid parameter, if not (but is still a string), then we will
+        # pass into category_name
+
+        uuid_pattern = re.compile(
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:141]: Either redefine regex as a class level variable OR don't use regex approach. Instead try to use plain **UUID()** constructor and catch **ValidationError** to determine if **category** is a valid uuid string.

File: tenable/io/v3/vm/tags/api.py [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+        }
+        payload = dict_clean(payload)
+
+        # let's validate payload using marshmallow schema
+        schema = TagCategorySchema()
+        payload = schema.dump(schema.load(payload))
+
+        return self._post('categories', json=payload)
+
+    def delete(self, *value_ids: UUID) -> None:
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:217]: Define ***value_ids** to **List[UUID]**.

File: tenable/io/v3/vm/tags/api.py [hunk 10]
```diff
@@ -0,0 +1,746 @@
...
+        '''
+        Deletes tag value(s).
+
+        :devportal:`tag: delete tag value <tags-delete-tag-value>`
+
+        Args:
+            *value_ids (str):
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:224]: Specify correct type hint information.

File: tenable/io/v3/vm/tags/api.py [hunk 11]
```diff
@@ -0,0 +1,746 @@
...
+                The unique identifier for the tag value to be deleted.
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:225]: Update docstring.
```suggestion
                The unique identifier for the tag values to be deleted.
```

File: tenable/io/v3/vm/tags/api.py [hunk 12]
```diff
@@ -0,0 +1,746 @@
...
+
+    def details(self, value_id: UUID) -> Dict:
+        '''
+        Retrieves the details for a specific tag category/value pair.
+
+        :devportal:`tag: tag details <tags-tag-value-details>`
+
+        Args:
+            value_id (str):
+                The unique identifier for the c/v pair
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:283]:
```suggestion
                The unique identifier for the category/value pair
```

File: tenable/io/v3/vm/tags/api.py [hunk 13]
```diff
@@ -0,0 +1,746 @@
...
+             current_domain_permissions: Optional[List] = None
+             ) -> Dict:
+        '''
+        Updates Tag category/value pair information.
+
+        :devportal:`tag: edit tag value <tags-update-tag-value>`
+
+        Args:
+            value_id (str):
+                The unique identifier for the c/v pair to be edited.
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:331]:
```suggestion
                The unique identifier for the category/value pair to be edited.
```

File: tenable/io/v3/vm/tags/api.py [hunk 14]
```diff
@@ -0,0 +1,746 @@
...
+        # this value will be incremented when permissions are updated
+        if 'version' in current['access_control']:
+            current_version = current['access_control']['version']
+        else:
+            current_version = 0
+
+        # version value must be incremented each time the
+        # permissions are updated
+        if payload['access_control'] != access_control:
+            payload['access_control']['version'] = current_version + 1
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:420]: Doesn't the concerned endpoint in the backend support partial updates and updates versions ? Calculating version in SDK seems to be heavyweight lifting and is not ideal in concurrent env, as it'll result in version conflicts during concurrent updates.

File: tenable/io/v3/vm/tags/api.py [hunk 15]
```diff
@@ -0,0 +1,746 @@
...
+                Examples:
+                    >>> [('field_name_1', 'asc'), ('field_name_2', 'desc')]
+            limit (int, optional):
+                Number of objects to be returned in each request.
+                Default and max_limit is 200.
+            next (str, optional):
+                The pagination token to use when requesting the next page of
+                results. This token is presented in the previous response.
+            return_resp (bool, optional):
+                If set to true, will override the default behavior to return
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:634]: Update docstring.

File: tenable/io/v3/vm/tags/api.py [hunk 16]
```diff
@@ -0,0 +1,746 @@
...
+                specific page of data.
+            return_csv (bool, optional):
+                If set to true, it will return the CSV response or
+                iterable (based on return_resp flag). Iterator returns all
+                rows in text/csv format for each call with row headers.
+        Returns:
+            iterable:
+                The iterable that handles the pagination for the job.
+            requests.Response:
+                If ``return_json`` was set to ``True``, then a response
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:645]: Update docstring.

File: tenable/io/v3/vm/tags/api.py [hunk 17]
```diff
@@ -0,0 +1,746 @@
...
+                that is to be returned.
+                Examples:
+                    >>> [('field_name_1', 'asc'), ('field_name_2', 'desc')]
+            limit (int, optional):
+                Number of objects to be returned in each request.
+                Default and max_limit is 200.
+            next (str, optional):
+                The pagination token to use when requesting the next page of
+                results. This token is presented in the previous response.
+            return_resp (bool, optional):
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:719]: Update docstring.

File: tenable/io/v3/vm/tags/api.py [hunk 18]
```diff
@@ -0,0 +1,746 @@
...
+                specific page of data.
+            return_csv (bool, optional):
+                If set to true, it will return the CSV response or
+                iterable (based on return_resp flag). Iterator returns all
+                rows in text/csv format for each call with row headers.
+        Returns:
+            iterable:
+                The iterable that handles the pagination for the job.
+            requests.Response:
+                If ``return_json`` was set to ``True``, then a response
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/api.py:731]: Update docstring.

File: tenable/io/v3/vm/tags/schema.py [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        required=True
+    )
+    description = fields.Str()
+
+
+class AssetTagSchema(Schema):
+    '''
+    Schema for asset tags methods
+    '''
+    action = fields.Str()
```
Comment (Mangusingh-te) [inline tenable/io/v3/vm/tags/schema.py:26]: Is there a master list of allowed values for **action** field ? If yes, then add validation to check if value belongs to master list.

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 20]
```diff
@@ -0,0 +1,830 @@
...
+        description=description
+    )
+
+    assert isinstance(res, dict)
+    assert res['name'] == name
+    assert res['description'] == description
+
+
+@responses.activate
+def test_delete_singluer_tag_value(api):
```
Comment (Mangusingh-te) [inline tests/io/v3/vm/tags/test_tags_api.py:160]:
```suggestion
def test_delete_singular_tag_value(api):
```

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 21]
```diff
@@ -0,0 +1,830 @@
...
+        responses.GET,
+        re.compile(f'{BASE_URL}/assets/{asset_id}/assignments'),
+        json=test_response
+    )
+
+    res = api.v3.vm.tags.list_asset_tag(asset_id=asset_id)
+
+    assert isinstance(res, dict)
+    for tag in res['tags']:
+        assert tag['asset_id'] == asset_id
```
Comment (Mangusingh-te) [inline tests/io/v3/vm/tags/test_tags_api.py:565]: Get rid of loop.

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 22]
```diff
@@ -0,0 +1,830 @@
...
+                'updated_by': 'szaros+us-2a@tenable.com',
+                'category_id': '4f3bd5fb-9f99-49c7-8deb-89f7596674d0',
+                'value': 'First Value',
+                'description': '',
+                'type': 'dynamic',
+                'category_name': 'Demo Tag',
+                'category_description': ''
+            }
+        ],
+        'pagination': {'total': 2, 'next': 'nextToken'}
```
Comment (Mangusingh-te) [inline tests/io/v3/vm/tags/test_tags_api.py:675]: **nextToken** is not needed as total records are 2 and they are already part of response.

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 23]
```diff
@@ -0,0 +1,830 @@
...
+        filter=filter,
+        sort=sort,
+        limit=200
+    )
+    assert isinstance(iterator, SearchIterator)
+
+    values_list = []
+    for item in iterator:
+        values_list.append(item)
+    assert len(values_list) == test_response['pagination']['total']
```
Comment (Mangusingh-te) [inline tests/io/v3/vm/tags/test_tags_api.py:732]: get rid of loop.

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 24]
```diff
@@ -0,0 +1,830 @@
...
+                'created_by': 'user@tenable.com',
+                'updated_at': '2021-10-19T12:06:17.529Z',
+                'updated_by': 'user@tenable.com',
+                'name': 'Audi',
+                'value_count': 1
+            }
+        ],
+        'pagination': {
+            'total': 1,
+            'next': 'nextToken'
```
Comment (Mangusingh-te) [inline tests/io/v3/vm/tags/test_tags_api.py:765]: nextToken is not needed as all records are already part of response.

File: tests/io/v3/vm/tags/test_tags_api.py [hunk 25]
```diff
@@ -0,0 +1,830 @@
...
+        filter=filter,
+        sort=sort,
+        limit=200
+    )
+    assert isinstance(iterator, SearchIterator)
+
+    categories_list = []
+    for item in iterator:
+        categories_list.append(item)
+    assert len(categories_list) == test_response['pagination']['total']
```
Comment (Mangusingh-te) [inline tests/io/v3/vm/tags/test_tags_api.py:820]: Get rid of loop.

**Review by Mangusingh-te — CHANGES_REQUESTED** [review]

**Comment by MitaliUpadhyay** [comment]
@alpesh-te, please close this PR as https://github.com/alpesh-te/pyTenable/pull/4 is created for these script changes.

**Pull request merged**